### ISSUE TITLE
PROPOSAL: Make the helpernode HA

### ISSUE THREAD

**Issue by christianh814** [redhat-cop/ocp4-helpernode#1]
# Proposal

Use Red Hat Cluster Suite  (`pacemaker`/`corosync`) to make the helpernode truly HA.

# Background

The helpernode was originally intended as a "teaching tool" to learn how to install OpenShift 4. I believe it's time to grow it beyond that and make it a viable tool to run in production.

**Comment by flickerfly** [comment]
Any ideas on how the file system would be sync'd between the cluster members? I'm thinking stuff like the pxe images share on the web server. Some of the other services would, of course, be able to handle this on their own like DNS.

**Comment by christianh814** [comment]
I think most services can be handled by `keepalived`. Data is a little difficult. Although I think you can do it with a replica volume using glusterfs.

**Comment by bpradipt** [comment]
May be we can phase this..
1. Phase-1: IP failover using keepalived. And manual copy of data to the two helper node instances. 
Since the helper node data is primarily static this might give us a good start to experiment and test the approach
2. Phase-2: Look at cluster filesystem like GFS2.

**Comment by flickerfly** [comment]
I like that. 

Maybe phase 1b, document what needs to be manually synced. I always have this idea in the back of my head I'm missing something when I have to make a list like that myself. :-) I'd probably start with a script to tar up the list on each member and store it on the opposing one for easy recovery, then iterate on that until a gfs2 plan is solidified.

**Comment by Prajyot-Parab** [comment]
**Experiment -**

_Environment used_ - **BareMetal Libvirt**

Created 2 Bastion domains (bastion & bastion_backup)
```
bastion - 192.186.88.2
bastion_backup - 192.168.88.3
VIP - 192.168.88.4
```

Executed necessary configs on both nodes (RH subscription, hostname setup, etc)

Ran Helpernode playbook on both nodes
- Used VIP as helper.ipaddr value
- Added entires in zonefile for bastion and bastion_backup

Executed necessary configs needed for keepalived on both nodes

**Testing -**

```
Node-1: Hosted Test.txt with content bastion-1 on bastion node
Node-2: Hosted Test.txt with content bastion-2 on bastion_backup node

pull with Node-1 ip on Node-1 -- o/p --> bastion-1
pull with Node-2 ip on Node-2 -- o/p --> bastion-2

pull with Node-1 ip on Node-2 -- o/p --> bastion-1
pull with Node-2 ip on Node-1 -- o/p --> bastion-1

pull with VIP on Node-1 -- o/p bastion-1 (as Node-1 serves as master)
pull with VIP on Node-2 -- o/p bastion-1 (as Node-1 serves as master)

Shutdown Node-1
pull with VIP on Node-2 -- o/p bastion-2 (as Node-2 serves as master now)
```


**Possible design changes needed:**

1. New var to enable HA - `high_availability: true`
2. In case of HA env `helper.ipaddr` should point to VIP (user concern while setting vars)
3. Create new var to be used only in case of HA
```
bastion:
  - name: "bastion"
    ipaddr: "192.186.88.2"
  - name: "bastion_backup"
    ipaddr: "192.186.88.3"
```

4. Make entries in Zonefile for both bastion nodes with their IPs (conditionally dependent on enabling HA var)
5. Ansible code for setting up keepalived ([reference](https://medium.com/@two.oes/openshift-4-in-an-air-gap-disconnected-environment-part-1-prerequisites-63f065e8a729))

**Comment by christianh814** [comment]
@Prajyot-Parab 

I have two thoughts so I'll break these out in sections

# Vars

I agree that changes need to be made. And I'm inclined to call this a new "x" release (in an x.y.z release); therefore I think we need to make "breaking changes" to what people have in their `vars.yaml` file (but not SO big :) )

I envisioned "breaking" the `helper` section and, as you suggested, create another section for HA. (I'm just suggesting these, if anyone has better names, please share):

```
high_availability:
  setup: true
  vip: 192.168.1.100
helpers:
  - name: "helper1"
    ipaddr: "192.168.1.77"
  - name: "helper2"
    ipaddr: "192.168.1.78"
```
The `high_availability.vip` will move between `helper1` and `helper2`.  The reason I really like your approach of having an array of helpers (or "bastions") is that you can easily add a 3rd helper. So if someone wants a 3 node HA setup, they can. 

If `high_availability.setup` is `false` then the playbook should...

1) Set up the first helper listed in the array and use that IP as the main IP (similar to what the playbook does now)
2) If a second helper exists in the array (and `high_availability.setup` is `false`), the playbook would simply ignore it (might be 

So really the only "breaking" change is that `helper` will get renamed (we don't have to rename it, just a thought), and become an array.

# Playbook run

So this playbook was meant to run on the helper itself, but now that we're adding another "helper" into the mix, then maybe we need change the instructions? I can see one of two solutions....

1) Still run the playbook on the fist helper, and use the vars.yaml file to take the second helper and dymanically add it to the inventory (difficult)
2) Instruct people to just add the proper entry in the `inventory` file. so maybe it'll look like this

```
[vmhost]
localhost ansible_connection=local

[helpers]
192.168.1.78
```

I think #2 is the easiest

# What is HA

So giving some thoughts...we should try to make everything run on the VIP and all the helpers as much as possible.

But for DNS...Should we just make secondary DNS servers that "slave" the data over?

I think that it might be less work to just use KeepAlived and have 1 master and 1+ standbys...which brings me to another point.

# Storage

I was giving some thought as to storage. What would be a good storage to use so that we don't have to replicate everything over? I was thinking of use glusterfs because of the fact that there's no "master" (they are all peers) and just setting up [a gluster replicated volume](https://docs.gluster.org/en/latest/Administrator%20Guide/Setting%20Up%20Volumes/#creating-replicated-volumes).

This way each helper in the `fstab` would point to itself. There is no failover needed for gluster. The `fstab` would look like this on helper1

```
192.168.1.77:helpernode	/opt/helpernode/	glusterfs	_netdev,backup-volfile-servers=192.168.1.78,acl 0 0
```

And on helper2

```
192.168.1.78:helpernode	/opt/helpernode/	glusterfs	_netdev,backup-volfile-servers=192.168.1.77,acl 0 0
```

Which brings me to mount...

# Mountpoint

Should we move all things that need to be HA into it's own directory/mount? This will require more work initially but I think it's worth exploring. It'll help us keep things in one dir that we can either sync/replicate without keeping track of which directories we'll need to sync.

# TLDR

That's probably more than you wanted, but I think if we think about where we want to go; it'll help us make better decisions up front.

For now, my thought is that setting up the vars like this:

```
high_availability:
  setup: true
  vip: 192.168.1.100
helpers:
  - name: "helper1"
    ipaddr: "192.168.1.77"
  - name: "helper2"
    ipaddr: "192.168.1.78"
```

is the most flexable.

**Comment by bpradipt** [comment]
> @Prajyot-Parab
> 
> I have many thoughts so I'll break these out in sections
> 
> # Vars
> I agree that changes need to be made. And I'm inclined to call this a new "x" release (in an x.y.z release); therefore I think we need to make "breaking changes" to what people have in their `vars.yaml` file (but not SO big :) )
> 
> I envisioned "breaking" the `helper` section and, as you suggested, create another section for HA. (I'm just suggesting these, if anyone has better names, please share):
> 
> ```
> high_availability:
>   setup: true
>   vip: 192.168.1.100
> helpers:
>   - name: "helper1"
>     ipaddr: "192.168.1.77"
>   - name: "helper2"
>     ipaddr: "192.168.1.78"
> ```
> 
> The `high_availability.vip` will move between `helper1` and `helper2`. The reason I really like your approach of having an array of helpers (or "bastions") is that you can easily add a 3rd helper. So if someone wants a 3 node HA setup, they can.
> 
> If `high_availability.setup` is `false` then the playbook should...
> 
> 1. Set up the first helper listed in the array and use that IP as the main IP (similar to what the playbook does now)
> 2. If a second helper exists in the array (and `high_availability.setup` is `false`), the playbook would simply ignore it (might be hard to do)
> 
> So really the only "breaking" change is that `helper` will get renamed (we don't have to rename it, just a thought), and become an array.
> 
> # Playbook run
> So this playbook was meant to run on the helper itself, but now that we're adding another "helper" into the mix, then maybe we need change the instructions? I can see one of two solutions....
> 
> 1. Still run the playbook on the fist helper, and use the vars.yaml file to take the second helper and dymanically add it to the inventory (difficult)
> 2. Instruct people to just add the proper entry in the `inventory` file. so maybe it'll look like this
> 
> ```
> [vmhost]
> localhost ansible_connection=local
> 
> [helpers]
> 192.168.1.78
> ```
> 
> I think option 2 is the easiest
> 
> # What is HA
> So giving some thoughts...we should try to make everything run on the VIP and all the helpers as much as possible.
> 
> But for DNS...Should we just make secondary DNS servers that "slave" the data over?
> 
> I think that it might be less work to just use KeepAlived and have 1 master and 1+ standbys...which brings me to another point.
> 
> # Storage
> I was giving some thought as to storage. What would be a good storage to use so that we don't have to replicate everything over? I was thinking of use glusterfs because of the fact that there's no "master" (they are all peers) and just setting up [a gluster replicated volume](https://docs.gluster.org/en/latest/Administrator%20Guide/Setting%20Up%20Volumes/#creating-replicated-volumes).
> 

@christianh814  should we look at GFS2 via RHEL resiliency storage add-on which  https://access.redhat.com/solutions/45930 ?

**Comment by Prajyot-Parab** [comment]
@christianh814 

### Vars

I would suggest something like this -

```
helper:
  name: "helper/vip/load-balancer-ip"  <--- name can be made generic 
  ipaddr: "192.168.1.100"
```
^^^ We keep helper var as it is, this would ensure we don't need to modify other things where `helper.ipaddr` is already mentioned. Also, it is already said in docs that `ipaddr` should be pointing to load balancing IP. We can update it to saying in case of HA this should be `VIP`.

**New Var**
```
high_availability:
  enable: true
bastions:
  - name: "bastion1"
    ipaddr: "192.168.1.77"
    state: MASTER
    priority: 100
  - name: "bastion2"
    ipaddr: "192.168.1.78"
    state: BACKUP
    priority: 98
```
^^^ Here incase of HA env, the user will need to set bastion details (this are merely needed to make DNS entries and generate respective keepalived conf files). By default, HA will be false.

The above approach will ensure we keep the present working state as it is and just develop on top of it without breaking anything. It will be also easier to understand for others.

```
{%if high_availability.enable %}
; Create entry for the bastion hosts
{% for b in bastions %}
{{ b.name | lower }}		IN	A	{{ b.ipaddr }}
{% endfor %}
;
{% endif %}
```
^^^Additional DNS entries code will be like this conditionally dependent on HA var

### Playbook run

`Instruct people to just add the proper entry in the inventory file.` -- I agree to this approach

Playbook can be executed on N bastions mentioned in inventory


### Storage
@bpradipt 
As per my understanding helpernode will ensure we have same behavior (set of services, binaries, etc) on both the nodes.

I had a quick question on what should be going in shared storage like

1. Post helper node / Pre-install artifacts?
 -- ignition files, install-config, manifest, etc.. <-- artifacts
2. Post-install artifacts?
^^^ Here I am trying to refer to artifacts actually needed for the cluster to function post-installation (this will the ones critical for maintaining HA env) like kubeconfig, auth dir, etc...
3. Both?

I may have missed some critical aspects here, please share your thoughts.

**Comment by Prajyot-Parab** [comment]
@christianh814 Do share your thoughts on ^^^

**Comment by christianh814** [comment]
> @christianh814
> ### Vars
> 
> I would suggest something like this -
> 
> ```
> helper:
>   name: "helper/vip/load-balancer-ip"  <--- name can be made generic 
>   ipaddr: "192.168.1.100"
> ```
> 
> ^^^ We keep helper var as it is, this would ensure we don't need to modify other things where `helper.ipaddr` is already mentioned. Also, it is already said in docs that `ipaddr` should be pointing to load balancing IP. We can update it saying in case of HA this should be `VIP`.
> 
> **New Var**
> 
> ```
> high_availability:
>   enable: true
> bastions:
>   - name: "bastion1"
>     ipaddr: "192.168.1.77"
>     state: MASTER
>     priority: 100
>   - name: "bastion2"
>     ipaddr: "192.168.1.78"
>     state: BACKUP
>     priority: 98
> ```
> 
> ^^^ Here incase of HA env, the user will need to set bastion details (this are merely needed to make DNS entries and generate respective keepalived conf files). By default, HA will be false.
> 
> The above approach will ensure we keep the present working state as it is and just develop on top of it without breaking anything. It will be also easier to understand for others.
> 
> ```
> {%if high_availability.enable %}
> ; Create entry for the bastion hosts
> {% for b in bastions %}
> {{ b.name | lower }}		IN	A	{{ b.ipaddr }}
> {% endfor %}
> ;
> {% endif %}
> ```
> 
> ^^^Additional DNS entries code will be like this conditionally dependent on HA var
> ### Playbook run
> 
> `Instruct people to just add the proper entry in the inventory file.` -- I agree to this approach
> 
> Playbook can be executed on N bastions mentioned in inventory

I think this is a good approach. This way it keeps the editing of the playbook to a minimum.


> ### Storage
> 
> @bpradipt
> As per my understanding helpernode will ensure we have same behavior (set of services, binaries, etc) on both the nodes.
> 
> I had a quick question on what should be going in shared storage like
> 
>     1. Post helper node / Pre-install artifacts?
>        -- ignition files, install-config, manifest, etc.. <-- artifacts
> 
>     2. Post-install artifacts?
>        ^^^ Here I am trying to refer to artifacts actually needed for the cluster to function post-installation (this will the ones critical for maintaining HA env) like kubeconfig, auth dir, etc...
> 
>     3. Both?


I think for now we should focus on the the services that need to have the storage mirrored somehow (glusterfs/GFS2/rsync).

**Comment by Prajyot-Parab** [comment]
Update -
1. Added new vars
```
high_availability:
  enable: true
bastions:
  - name: "bastion1"
    ipaddr: "192.168.1.77"
    state: MASTER
    priority: 100
  - name: "bastion2"
    ipaddr: "192.168.1.78"
    state: BACKUP
    priority: 98
```
```
Note:
1. helper.ipaddr should point to VIP/load balancer VIP in case of high_availability and multiple bastion nodes
2. It is mandatory to define bastions vars in case of high_availability to ensure keepalived setup is actually done (checks added in code for the same)
3. Can work with 1 to N number of bastions (theoretically)
```

2. Code changes to update DNS
3. Ansible tasks for setting up keepalived
4. Single `keepalived.conf` file template for (MASTER, BACKUP)  nodes

5. Playbook run via updating inventory with bastion details
- Clone and run from one of the bastion nodes
- Clone and run from remote machine

@christianh814  I have tested the above-mentioned changes and features in my local environment (worked as expected) and will do some more testing for the same.

@bpradipt @christianh814 Now we can focus on the second phase i.e shared storage perspective, maybe decide on what we plan to use to achieve the same.

**Comment by walidshaari** [comment]
I would say keep it simple for storage, no need to add complexity and dependency, simple rsync to sync the data if you like.
